File: examples/mqtt_publish.mjs
import mqtt from "mqtt";

const host = process.env.MQTT_HOST ?? "broker.hivemq.com";
const port = Number(process.env.MQTT_PORT ?? "1883");
const topicPrefix = (process.env.MQTT_TOPIC_PREFIX ?? "electrogpt/telemetry").replace(/\/+$/, "");
const channel = process.env.MQTT_CHANNEL ?? "atelier-ligne-1";
const username = process.env.MQTT_USERNAME ?? "";
const password = process.env.MQTT_PASSWORD ?? "";
const count = Number(process.env.MQTT_COUNT ?? "20");
const intervalMs = Number(process.env.MQTT_INTERVAL_MS ?? "1000");

const topic = `${topicPrefix}/${channel}`;
const client = mqtt.connect(`mqtt://${host}:${port}`, {
  username: username || undefined,
  password: password || undefined,
  keepalive: 60,
});

function buildPayload(sequence) {
  return {
    channel,
    values: {
      temperature_c: Number((45 + 3.5 * Math.sin(sequence / 3)).toFixed(3)),
      current_a: Number((18 + 2.2 * Math.cos(sequence / 4)).toFixed(3)),
      voltage_v: Number((398 + 4 * Math.sin(sequence / 5)).toFixed(3)),
      power_factor: Number((0.9 + 0.03 * Math.sin(sequence / 6)).toFixed(4)),
    },
    metadata: {
      publisher: "node-example",
      sequence,
      machine: "transformer-1",
    },
  };
}

client.on("connect", () => {
  console.log(`Publishing to topic: ${topic}`);
  let sequence = 1;

  const timer = setInterval(() => {
    const payload = buildPayload(sequence);
    const body = JSON.stringify(payload);
    client.publish(topic, body, { qos: 0, retain: false });
    console.log(`[${sequence}/${count}] ${body}`);

    sequence += 1;
    if (sequence > count) {
      clearInterval(timer);
      setTimeout(() => client.end(), 300);
    }
  }, Math.max(intervalMs, 50));
});

client.on("error", (error) => {
  console.error("MQTT error:", error.message);
  client.end(true);
});

File: examples/mqtt_publish.py
import argparse
import json
import math
import os
import time

import paho.mqtt.client as mqtt


def create_client() -> mqtt.Client:
    if hasattr(mqtt, "CallbackAPIVersion"):
        return mqtt.Client(mqtt.CallbackAPIVersion.VERSION2, client_id="electrogpt-python-publisher")
    return mqtt.Client(client_id="electrogpt-python-publisher")


def build_payload(channel: str, sequence: int) -> dict:
    temperature = 45.0 + 3.5 * math.sin(sequence / 3.0)
    current = 18.0 + 2.2 * math.cos(sequence / 4.0)
    voltage = 398.0 + 4.0 * math.sin(sequence / 5.0)
    power_factor = 0.9 + 0.03 * math.sin(sequence / 6.0)
    return {
        "channel": channel,
        "values": {
            "temperature_c": round(temperature, 3),
            "current_a": round(current, 3),
            "voltage_v": round(voltage, 3),
            "power_factor": round(power_factor, 4),
        },
        "metadata": {
            "publisher": "python-example",
            "sequence": sequence,
            "machine": "transformer-1",
        },
    }


def main() -> None:
    parser = argparse.ArgumentParser(description="Publie des donnees MQTT pour le dashboard live ElectroGPT.")
    parser.add_argument("--host", default=os.getenv("MQTT_HOST", "broker.hivemq.com"))
    parser.add_argument("--port", type=int, default=int(os.getenv("MQTT_PORT", "1883")))
    parser.add_argument("--topic-prefix", default=os.getenv("MQTT_TOPIC_PREFIX", "electrogpt/telemetry"))
    parser.add_argument("--channel", default=os.getenv("MQTT_CHANNEL", "atelier-ligne-1"))
    parser.add_argument("--count", type=int, default=20, help="Nombre de messages a publier")
    parser.add_argument("--interval", type=float, default=1.0, help="Intervalle entre deux publications en secondes")
    parser.add_argument("--username", default=os.getenv("MQTT_USERNAME", ""))
    parser.add_argument("--password", default=os.getenv("MQTT_PASSWORD", ""))
    args = parser.parse_args()

    topic = f"{args.topic_prefix.strip('/')}/{args.channel}"
    client = create_client()
    if args.username:
        client.username_pw_set(args.username, args.password or None)

    client.connect(args.host, args.port, 60)
    client.loop_start()

    print(f"Publishing to topic: {topic}")
    try:
        for sequence in range(1, args.count + 1):
            payload = build_payload(args.channel, sequence)
            body = json.dumps(payload)
            client.publish(topic, body, qos=0, retain=False)
            print(f"[{sequence}/{args.count}] {body}")
            time.sleep(max(args.interval, 0.05))
    finally:
        client.loop_stop()
        client.disconnect()


if __name__ == "__main__":
    main()
